src/base/TypeChecker.ml

```diff
@@ -798,6 +798,36 @@ module ScillaTypechecker (SR : Rep) (ER : Rep) = struct
             @@ add_stmt_to_stmts_env_gas
                  (TypedSyntax.CallProc (p, typed_args), rep)
                  checked_stmts remaining_gas
+        | ListIter (l, p) -> (
+            let%bind lt =
+              mark_error_as_type_error remaining_gas
+              @@ TEnv.resolveT env.pure (get_id l) ~lopt:(Some (get_rep l))
+            in
+            let l_type = rr_typ lt in
+            match
+              List.Assoc.find env.procedures ~equal:String.( = ) (get_id p)
+            with
+            | Some [ arg_typ ] ->
+                let%bind _ =
+                  wrap_type_serr stmt
+                  @@ mark_error_as_type_error remaining_gas
+                  (* The procedure accepts an element of l. *)
+                  @@ assert_type_equiv (list_typ arg_typ) l_type.tp
```

Comment:
This line restricts `forall` to work on lists, but I think it should work for nats as well. In general, it should work on any inductive datatype.

Reply:
Resolving this one, as we have decided to only support lists for now.